### ISSUE TITLE
REngine.GetInstance() method goes in infinite loop after building application.

### ISSUE THREAD

**Issue by ravirupeliya** [jmp75/rdotnet#39]
I am using R.Net with my web application. I have hosted site in IIS 7 and written below simple code on click event of a button

```
protected void btnGetData_Click(object sender, EventArgs e)
{
    //Get instance of REngine
    REngine.SetEnvironmentVariables();
    engine = REngine.GetInstance();
}
```
My problem is, above code works fine for first time, but when I rebuild my solution then again run the same page, then the line `engine = REngine.GetInstance();` goes into infinite loop.
I have also tried using `engine.Dispose()` method, but then it gives below error.

> The single REngine instance has already been disposed of (i.e. shut down). Multiple engine restart is not possible.

Can any body please help me out?.

**Comment by hsnarula** [comment]
Hi 

Is there any solution to this problem. We are facing the same issue.

Regards
Harvinder

**Comment by ravirupeliya** [comment]
Nope. But I managed it by restarting IIS or application pool after building application !!!

**Comment by rknoll** [comment]
Hi, I just spent 3 days with this and believe that I finally managed to find the cause..

The problem is, that by only re-running the application, **IIS keeps the worker process alive** - and with that, the `R.dll` stays in memory and you simply can't re-initialize it. So the first run will work, but after that you'll have to restart / recycle the application pool in order to stop the process.

As a workaround, I subclassed the `REngine` and modified the `Initialize` method to allow me to still call `Rf_initialize_R` and `R_SetParams_Windows`, but not `setup_Rmainloop`, as this is where the R code jumps into an endless loop (if it has been called already before), and there is no way to stop the Thread..
The way I check if I have to do that is by simply looking through the loaded modules in `Process.GetCurrentProcess().Modules` and check if the `R.dll` is in there. I've also modified the `Dispose` method to not run any R finalizers.

This all seems like a big messy workaround, and I hope someone with a better understanding of the internals of R and R.NET will help us find a better one 😃 .
